docs/rfcs/0011-customizable-store-dispatcher.md

```diff
@@ -0,0 +1,71 @@
+---
+layout: page
+title: Allow customisation of the Store dispatcher
+permalink: /rfc/0011-customizable-store-dispatcher
+---
+
+* Start date: 2024-01-29
+* RFC PR: [#5353](https://github.com/mozilla-mobile/firefox-android/pull/5353)
+
+## Summary
+
+In most applications of the `Store`, it is preferable to have reducers perform work on the main thread. This helps to simplify the levels of async tasks that need to be synchronised with other async tasks that are already on the main thread.
+
+## Motivation
+
+Android embedders use the main thread for UI, user-facing, or touch handling work. For example, notifying UI components when IO from storage layers have completed, an engine's tasks that can happen on a separate thread, or global-level state updates for different components to observe.
+
+When components dispatch actions to the store, these actions are performed on an independant single thread dispatcher in the `Store` to avoid overloading the main thread with heavy work that might be performed during the `reduce` or in a `Middleware`. In practice, these actions have been short and fast so they do not cause overhead. Work done in a `Middleware` that can be slow, like I/O,  is put onto separate Dispatchers. The performance optimization to switch to a `Store` thread, requires that components which are always run on the main thread, to ensure synchronisation is now kept between the main thread and the store thread for observers of the `State`.
```

Comment:
> In practice, these actions have been short and fast so they do not cause overhead.

Are there any profiles or measurements that have shown this to be true?

Reply:
This is only anecdotal evidence because it really depends on the specific store, so it's not worth adding to the RFC: I ran our performance tests against a prototype patch where I changed the Fenix AppStore to use the main dispatcher and there was only a small 33ms perf win (2199.0-2166.0) using a debug build.

If we start to switch over stores to use the main dispatchers we can run tests for those to see if they are quicker and no longer doing work on the wrong thread.

```
==> patch-pixel6a-store-mainthread-0.txt <==
{'max': 2256.0,
 'mean': 2104.9,
 'median': 2166.0,
 'min': 1514.0,
 'replicate_count': 10,
 'replicates': [2097.0, 1514.0, 2067.0, 2157.0, 2163.0, 2169.0, 2183.0, 2191.0,
                2252.0, 2256.0]}

==> patch-pixel6a-store-mainthread-pre-0.txt <==
{'max': 2929.0,
 'mean': 2220.9,
 'median': 2199.0,
 'min': 1613.0,
 'replicate_count': 10,
 'replicates': [2929.0, 1613.0, 2098.0, 2186.0, 2106.0, 2154.0, 2249.0, 2212.0,
                2317.0, 2345.0]}

```